### PR TITLE
fix(geometry): finite float16 gradients in axis_angle_to_rotation_matrix near the identity

### PR DESCRIPTION
**Pull request by mohansree14**
## What does this pull request do?

`axis_angle_to_rotation_matrix` returns `nan` gradients in `float16` at the identity and for any small rotation whose `theta2` underflows:

```python
aa = torch.tensor([[0.0, 0.0, 0.0], [1e-4, 0.0, 0.0]], dtype=torch.float16, requires_grad=True)
axis_angle_to_rotation_matrix(aa).sum().backward()
aa.grad  # main: all nan; bfloat16 / float32: finite
```

Small angles take the Taylor branch, but `torch.where(mask, rot_normal, rot_taylor)` still backpropagates through the discarded Rodrigues branch. That branch computed `sqrt(theta2.clamp(min=1e-12))`, and `1e-12` is `0` in `float16`, so its backward differentiated `sqrt(0)` and the `inf` turned into `nan` in the `where`. This is the pattern AGENTS.md describes from #4229: a floor below the dtype's smallest subnormal is no guard.

The fix substitutes a safe argument instead of flooring it:

```python
mask = theta2 > 1e-6
safe_theta2 = torch.where(mask, theta2, torch.ones_like(theta2))
rot_normal = _compute_rotation_matrix(axis_angle, safe_theta2)
```

Rows that take the Rodrigues branch see the same `theta2` as before (`theta2 > 1e-6` is well above the old floor), so forward values are unchanged in every dtype. The clamp is removed because nothing reaches that `sqrt` at zero anymore.

## Related issue or discussion

No separate issue. This is a follow-up to the float16 floor audit in #4229, for a function that audit did not cover.

## What did you check?

- New `TestAngleAxisToRotationMatrix::test_convention_gradient_is_finite_near_the_identity` (identity, `1e-4` and a mixed small vector; asserts finite gradients and zero gradient at the identity). It fails on `main` in `float16` (`nan`) and passes in `float16`, `bfloat16`, `float32` and `float64` with the fix.
- `pytest tests/geometry/test_conversions.py tests/geometry/liegroup`: 804 passed, 1 skipped, 7 xfailed (float32).
- `pytest tests/geometry --dtype=float16` and `--dtype=bfloat16` with `--xfail-known-failures --known-failure-profile=cpu-<dtype>`, on `main` and on this branch: the failing sets are identical except for the new test, which is fixed. The failures they share come from my local environment (Windows, torch 2.14.0 CPU) and are unrelated to this function.
- Doctests in `kornia/geometry/conversions.py`: 34 passed.
- `ruff check` and `ruff format --check` (0.16.7) clean on the changed files.

Not run: full `pre-commit-all`, `ty`, the torch 2.5.1 leg, CUDA or MPS.

## How did AI help?

Claude (Opus 5) found the underflow, wrote the fix and the test, ran the checks above including the before/after half-precision comparison, and drafted this description.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
## Review — #4613, head `e978b935`

This is a correct, well-scoped fix and the analysis in your description holds up under execution — including the part that would have been easy to get wrong. Merge onto `ecbcad561` is clean (`d7b419339`). Two P2s before merge.

### The mechanism is real, and it is not a torch-2.14 artifact

That was the thing worth checking, since AGENTS.md's `clamp` trap is version-bound. It is not one here:

| | float16 grad on main | on this branch |
| --- | --- | --- |
| torch 2.14.0 | `[nan, nan, nan]` at identity and at `1e-4` | `[0., 0., 0.]` / `[-2.0003e-04, 1.0002e-04, 1.0002e-04]` |
| torch 2.5.1 | same `nan` | same finite |

bfloat16/float32/float64 are finite on main, because `1e-12` is representable there and the floor works. The nan appears on **both** torch legs because the `where`'s zero upstream gradient multiplies `sqrt'(0) = inf` *before* the clamp backward runs — I confirmed the underlying version difference separately (`torch.zeros(1).clamp(min=0).sum().backward()` → grad `1.0` on 2.5.1, `0.0` on 2.14), and it does not rescue main either way.

### The gradients are correct, not merely finite

Independent float64 reference = `torch.matrix_exp(skew(v))` autograd on the float16-rounded inputs, cross-checked against central finite differences (the two agree to 3.6e-10). Over the **1932 rows that were `nan` on main**: max abs error **2.98e-07**, median 2.2e-16, with `|ref|` up to 6.04e-04.

### Forward is bit-identical — executed, not argued

6020 axis-angle inputs (4000 log-spaced magnitudes 1e-8..10, 20 hand-picked boundary values including `1e-3`, `1.0000001e-3`, `9.999999e-4`, and 2000 `randn`), `torch.equal` main vs branch:

| dtype | forward differing | grad differing |
| --- | --- | --- |
| float64 | **0 / 54180** | 0 / 18060 |
| float32 | **0 / 54180** | 0 / 18060 |
| bfloat16 | **0 / 54180** | 0 / 18060 |
| float16 | **0 / 54180** | 5796 / 18060 — exactly the 1932 `nan` rows × 3 |

Identical table on **torch 2.5.1**, and on **MPS** (3005 inputs, f32+f16: forward `torch.equal` True both dtypes; f16 grad differs only in the 4707 previously-`nan` elements).

**The branch boundary is unchanged.** `mask = theta2 > 1e-6` is byte-for-byte main's predicate, just hoisted out of the `[..., None, None]` index — I compared the mask tensor itself with `torch.equal`: True in all four dtypes (n_true 3782/3783/3781/3776). So no behaviour change and no breaking note is owed.

**No suite regression.** `tests/geometry/test_conversions.py tests/geometry/liegroup` at float32: 804 passed, 1 skipped, 7 xfailed. At float16: 60 failed, 739 passed — and the failure node list is **byte-identical to main's** (`diff` empty, 60 vs 60). Dynamo leg green (12 passed), doctests 34 passed, pinned `ruff@0.16.7` clean.

---

## [P2] The new pin is finiteness-only — it pins no gradient value

`rotation_matrix = 2.0 * torch.where(...)` — every gradient doubled — **survives** the new test in all four dtypes, because `isfinite` is scale-invariant and the one value assert is `grad[0] == 0`, which `2*0` also satisfies.

In half precision that assert is close to vacuous: `BaseTester`'s float16 tolerance is `atol=1e-3` and every gradient in the test is O(1e-4), so any finite value passes.

This is measured and passes f16/bf16/f32/f64 while **killing** the 2.0× mutant in both f16 and f32:

```python
self.assert_close(
    axis_angle.grad[1:],
    torch.tensor([[-2e-4, 1e-4, 1e-4], [7e-4, -8e-4, 1e-4]], device=device, dtype=dtype),
    rtol=1e-2, atol=1e-6,
)
```

(values are the float64 `matrix_exp` reference; largest observed deviation is bfloat16 at 1.9e-6, inside `atol + rtol*|ref|` = 9e-6.)

## [P2] Missing the Convention bullet that both sibling fixes established

The two functions already fixed with this exact substitution each carry one. `quaternion_to_axis_angle` (`conversions.py:912`):

> differentiable at the identity quaternion. The `sqrt` whose derivative is unbounded there never sees its zero radicand, so the gradient is the analytic limit of the surrounding map rather than `nan`. The guard is elementwise, and away from the identity it moves no forward bit.

`axis_angle_to_quaternion` (`conversions.py:1233`) has the same bullet worded for the zero rotation. `axis_angle_to_rotation_matrix` now has the same guarantee and no bullet — I grepped its whole docstring for `differenti|gradient|nan`: no hit. It already has a `Convention:` block, so there is a home for it. My sweep above supplies the "moves no forward bit" clause as an executed fact.

## [P3] The third test row does not discriminate

`[0., 5e-4, 2e-4]` has `theta2 = 2.9e-7`, a representable float16 subnormal, so `sqrt` never sees 0 and **main already returns the correct gradient there** (`[7.0000e-04, -8.0013e-04, 1.0014e-04]`, matching float32). The full-revert mutation fails only on the identity and `1e-4` rows. Harmless, but the comment implies all three rows underflow.

## [P3] The same defect class survives in siblings — but don't grow this PR

float16 only (float32/float64 clean), all going through `safe_zero_division`'s `clamp(min=eps)` / `sqrt(x + eps)` in the same file:

- `rotation_matrix_to_quaternion(eye(3))` → grad `[nan, -0.25, 0.25, 0.25]`; also at 180°, `diag(1,-1,-1)` → `[nan, 0.25, 0.25, 0.25]`
- `rotation_matrix_to_axis_angle(eye(3))` → `[nan, -0.5, 0.5, 0.5]`
- `normalize_quaternion(zeros(4))` → `inf`

`quaternion_to_axis_angle`, `axis_angle_to_quaternion`, `quaternion_to_rotation_matrix`, `So3.exp` and `Se3.exp` are all clean — they already use the substitution. The first of these is #3981's territory (its outstanding ask is exactly "version-independent gradient guard, half coverage"); I will make sure the other two are tracked. Out of scope here.

## Mutation testing

Harness sanity first: a `+1.0` constant offset survived — correctly, since it does not change a gradient — so I used a scaling mutant to prove the harness can go red.

| mutation | result |
| --- | --- |
| `2.0 * torch.where(...)` | **SURVIVED** → the first P2 |
| full revert to main (`sqrt(theta2.clamp(min=1e-12))`) | **killed** — 1 failed, float16 only; confirms your "fails on main in float16" |
| `safe_theta2 = theta2` | **killed** — 4 failed, all dtypes |
| keep substitution, restore `clamp(min=1e-12)` inside | survived — correctly, a true no-op once the substitution binds |
| `torch.ones_like` → `torch.zeros_like` | **killed** — 4 failed |
| mask widened to `theta2 >= 0` | **killed** — 4 failed |
| floor-only non-fix `theta2.clamp(min=1e-4)` | survived the pin and the whole suite — an honest limit of the pin, recorded below |
| `rotation_matrix = rot_normal` (always Rodrigues) | survived the pin; **killed by the existing suite** (2 failed f32, 62 f16) |
| `rotation_matrix = rot_taylor` (always Taylor) | survived the pin; killed by existing float32 tests |

On the floor-only survivor: `clamp(min=1e-4)` is float16-representable, so it also removes the `nan`. It does perturb the Rodrigues branch for `1e-6 < theta2 < 1e-4`, but only at ~1.6e-5 relative (`R[2,1]` at `v=(2e-3,0,0)`: 0.00199996680 vs 0.00199999870), under every tolerance in the suite. That is a limit of the pin, not a defect in your fix — **your** fix is bit-exact where that one is not, which is precisely why the substitution is the right pattern.

### Also verified

Changelog fragment `changelog.d/4613.fixed.md` is correctly named and tells the truth (finite gradients near the identity in float16; it claims no forward change, and none happens). No convention-pin debt — the PR closes no issue, `grep -rnE "#4613|issues/4613" kornia/ tests/` on the merge tree is empty, and no `test_wart_*` pin exists for this function on main. No new public names, so no `docs/source/*.rst` entry owed.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Review by ducha-aiki — APPROVED** [review]
Review of `77ebf9913ac2fc4bf6f0723c691debc22e00e284` — **APPROVE**.

Both requested changes are complete. The regression now pins the independently derived gradient values with a tight tolerance, so a finite but wrong gradient no longer passes, and the public Convention block records the identity-gradient guarantee. The production substitution remains forward-bit-identical away from the guarded branch and removes the float16 `sqrt(0)` backward path.

Validation: the focused regression passed for CPU float16, bfloat16, float32, and float64 (4 passed); the full conversion suite passed for CPU float32/float64 (651 passed, 2 skipped, 11 expected failures). The merge tree against current `main` is clean and all 41 reported checks are green. No actionable findings remain.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).
